blog/content/post/2020-08-26-coreboot-riscv.md

```diff
@@ -0,0 +1,267 @@
+cover: /covers/coreboot-logo.svg
+author: wojciech.niewiadomski
+layout: post
+published: true
+date: 2020-08-26
+archives: "2020"
+
+tags:
+  - Firmware
+  - coreboot
+  - RISC-V
+categories:
+  - Firmware
+  - Miscellaneous
+
+---
+
+Your post content
+
+# Building coreboot for RISC-V and running it in Qemu
+
+#### 1. Install required tools and libraries
+
+On Debian:
+
+```sh
+$ sudo apt-get install -y bison build-essential curl flex git gnat libncurses5-dev m4 zlib1g-dev qemu-system
+```
+On Arch Linux:
+
+```sh
+$ sudo pacman -S base-devel curl git gcc-ada ncurses zlib
+```
+
+On Fedora:
+
+```sh
+$ sudo dnf install git make gcc-gnat flex bison xz bzip2 gcc g++ ncurses-devel wget zlib-devel
+```
+
+#### 2. Download coreboot source tree
+
+```sh
+$ git clone https://review.coreboot.org/coreboot
+$ cd coreboot
+```
+
+#### 3. Build coreboot toolchain
+
+You can use CPUS= to specify the number of threads you want to use for building
+toolchain.
+
+```sh
+$ make crossgcc-riscv CPUS=4
+```
+> NOTE: This will take a while
+
+At the end of the process, you should get the info that your cross toolchain is
+built
+
+```
+You can now run riscv64-elf cross GCC from /your/path/to/coreboot/util/crossgcc/xgcc.
+```
+
+#### 4. Configure payload
+
+> Unfortunately using demonstration payloads such as `coreinfo` or `tint` is not possible as they use `libpayload` library which does not support RISC-V architecture yet. However, there is a [WIP branch](https://review.coreboot.org/c/coreboot/+/31356) working on adding initial support for RISC-V you can check out.
+
+
+As an example payload I will use SeaBIOS, which you can later use to boot the operating system.
+
+
+Firstly you need to clone the repository.
+
+```sh
+$ git clone https://review.coreboot.org/seabios.git
+$ cd seabios
+```
+
+Now configure the `menuconfig`
+
+```sh
+$ make menuconfig
+```
+
+Inside `menuconfig` follow these steps:
+
+```
+   select 'General Features' menu
+   select 'Build for coreboot' in 'Build Target'
+   select < Exit >
+   select < Exit >
+   select < Yes >
+```
+
+```sh
+$ make
+```
+
+You should get the output file in `out/bios.bin.elf`
+
+#### 5. Configure the build
+
+Configure your mainboard
+
+```sh
+$ make menuconfig
+```
+
+Inside `menuconfig` follow these steps:
+
+```
+   select 'Mainboard' menu
+   select '(Emulation)' in 'Mainboard vendor'
+   select 'QEMU RISC-V rv64' in 'Mainboard model'
+   select < Exit >
+   select `Payload` menu
+   select '(An ELF executable payload)' in 'Add a payload'
+   in 'Payload path and filename' add path 'seabios/out/bios.bin.elf'
+   select < Exit >
+   select < Exit >
+   select < Yes >
+```
+
+(Optionally) You can check your configuration by these commands:
+
+```sh
+$ make savedefconfig
+$ cat defconfig
+```
+
+The output should look like this:
+
+```
+CONFIG_BOARD_EMULATION_QEMU_RISCV_RV64=y
+CONFIG_PAYLOAD_ELF=y
+CONFIG_PAYLOAD_FILE="seabios/out/bios.bin.elf"
+```
+
+#### 6. Build coreboot
+
```

Comment:
avoid empty lines (unless they are specifically needed in this case for some reason?)